src/webrtc/gcc.md

```diff
@@ -0,0 +1,159 @@
+or to reencode the video on its own.
+The former is problematic in sessions with more than 2 participants as there might be e.g. only one host that
+cannot handle high quality video but the server will send all of the participants lower resolution.
+The latter requires a lot of resources and is hard to be achieved in the realtime.
+This leads us to the simulcast where the sender sends multiple qualities in multiple FPS and server decides
+which quality should be sent to which participant.
+There are still some bitrate levels so the server cannot lower its rate to any bitrate level but in most cases
+there is always some quality that will fit into the receiver network condition.
+
+As stated before, bandwidth estimation depends on the amout of data that is sent therefore, if the sever sends
+little data the estimation will be low.
+To recover from this situation, the server has to increase its sending rate which leads us to the connection probing.
+
+## Connection Probing
+
+When the server wants to recover from sending low resolution to some peer, it has to generate some additional traffic
+to see if the network can handle more data.
+
+Unfortuantely, the connection probing is not standardized so every SFU might do this in a different way.
+
+The questions we need to answer are:
+1. what data use for probing
+1. how much to probe
+1. how often to probe
+1. how long to probe
+
+### 1. What data use for probing
+
+The naive implementation would be to switch to the higher layer every `x` seconds and observe the network.
+This however, may lead to periodic video rebuffering and lower Quality of Experience (QoE).
+
+Another solution is to generate traffic by sending RTP padding packets.
+In this case, server has full control over the amout of data being sent and therefore, it can probe with 
+different aggresivness.
+
+The problem with padding packets is that they contain only zeros (except the last byte) so we end up sending
+a lot of garbage data, useless for the client. 
+Fortunatelly, there is a solution which is called RTX.
+RTX is a shortcut for retransmissions and can be used for probing instead of padding packets.
+How does it work?
+Instead of sending padding packets we simply retranssmit some packets via dedicated RTP stream.
+This way we have a full control over probing rate and we send data that is useful for the client.
+That's also the way libwebrtc chooses when RTX extension is enabled.
+
+#### Probing with padding packets
+
+In our webrtc implementation, we started with padding packets as they are easier for implementing and give
+us pretty good results.
+Here are some tips important when using padding packets:
+* the `P` bit in the RTP header has to be set to `1`.
+It indicates that RTP packet contains padding at the end of a payload.  
+* maximal size of padding only payload is 255.
+The limitation comes from the last byte. When the `P` flag is set, the last byte in payload denotes number of 
+zeros that should be stripped out.
```

Comment:
It just occurred to me. Wouldn't we want to introduce a concept of padding packets separately from probing?

Reply:
Good point. I will introduce this term befor going into tips